src/nextjournal/clerk/hashing.clj

```diff
@@ -199,40 +200,53 @@
 #_(parse-file {:markdown? true} "notebooks/rule_30.clj")
 #_(parse-file "notebooks/src/demo/lib.cljc")
 
+(defn- circular-dependency-error? [e]
+  (-> e ex-data :reason #{::dep/circular-dependency}))
+
+(defn- analyze-circular-dependency [state var form dep {:keys [node dependency]}]
+  (let [rec-form (concat '(do) [form (get-in state [:var->hash dependency :form])])
+        rec-var (symbol (str var "+" dep))]
+    (-> state
+        (update :graph #(-> %
+                            (dep/remove-edge dependency node)
+                            (dep/depend var rec-var)
+                            (dep/depend dep rec-var)))
+        (assoc-in [:var->hash rec-var :form] rec-form))))
+
+(defn- analyze-deps [var form {:as state :keys [graph]} dep]
+  (try (assoc state :graph (dep/depend graph (if var var form) dep))
+       (catch Exception e
+         (when-not (circular-dependency-error? e)
+           (throw e))
+         (analyze-circular-dependency state var form dep (ex-data e)))))
```

Comment:
none of the in-line repl examples hit this code path so these changes aren't verified :grimacing: maybe we can extend the test coverage before going forward with this change